attribution/api_attribution.py

```diff
@@ -46,11 +53,200 @@ def __init__(
+        """
+        llm_input = LLMInput(
+            input_string=original_input, tokenizer=self.tokenizer, unit_definition=unit_definition
+        )
+
+        original_output = await self.get_chat_completion(llm_input.input_string)
+        if logger:
+            logger.start_experiment(
+                original_input,
+                original_output.message.content,
+                perturbation_strategy,
+                unit_definition,
+            )
+
+        # Defining boolean masks via a sliding window to split the input into chunks
+        unit_count = len(llm_input.unit_tokens)
+        masks = get_masks(unit_count, init_chunk_size, stride)
+
+        comparator = np.abs if use_absolute_attribution else lambda x: x
+        cumulative_unit_attribution = np.zeros(unit_count)
+        total_llm_calls = 1
+        stage = 0
+
+        # Main loop for hierarchical perturbation, run until masks cannot be further subdivided
+        while masks:
+            print(f"Stage {stage}: making {len(masks)} perturbations")
+            # Define perturbations for each mask
+            perturbations: list[PerturbedLLMInput] = []
+            for mask in masks:
+                perturbations.append(
+                    PerturbedLLMInput(
+                        original=llm_input,
+                        strategy=perturbation_strategy,
+                        perturb_unit_ids=np.where(mask)[0].tolist(),
+                    )
+                )
+
+            if verbose:
+                print("Masked out tokens/words:")
+                print(*[[perturbation.masked_string] for perturbation in perturbations], sep="\n")
+
+            # Wait for the perturbed results
+            outputs = await self.get_multiple_completions(
+                [perturbation.perturbed_string for perturbation in perturbations]
+            )
+            total_llm_calls += len(outputs)
+
+            # Calculate attribution scores for each perturbation
+            chunk_scores = []
+            unit_attribution = np.full((len(masks), unit_count), np.nan)
+
+            for i, (perturbation, output, mask) in enumerate(zip(perturbations, outputs, masks)):
+                for strategy in attribution_strategies:
+                    # Logging each attribution strategy metric
+                    attribution_scores, norm_attribution_scores = self._get_scores(
+                        perturbation=perturbation,
+                        perturbed_output=output,
+                        original_output=original_output,
+                        attribution_strategy=strategy,
+                        chunksize=sum(mask),
+                        ignore_output_token_location=ignore_output_token_location,
+                        logger=logger,
+                        depth=stage,
+                    )
+
+                    # For threshold calculation and comparison we only use the first attribution strategy metric
```

Comment:
Not sure I like this. It's opaque to the user and potentially confounding – using one attribution strategy for thresholding and another for score assignation, and labelling the results with the latter – and this depending on the ordering of the strategy list. I think we should default to all algorithms using a single attribution strategy, instead of taking a list?

Reply:
Yes! I wanted to keep multiple attribution strategies because that was the original implementation, but I think it makes things more clunky, especially thresholding where I agree it's a bit arbitrary. I will update both `compute_attributions` and `hierarchical_perturbation` to take a single strategy if we're happy with that?

If we stick with two I guess we could have another argument to specify which is being used for thresholding, but my preference would be one strategy.